htlcswitch/link.go

```diff
@@ -609,47 +701,107 @@ func (l *channelLink) IsDraining(linkDirection LinkDirection) bool {
 // channel state reaches zero htlcs. This hook will only ever be called once.
 // This function returns a FlushId that can be used to preemptively remove this
 // hook.
-func (l *channelLink) OnFlushedOnce(func()) fn.Option[FlushHookID] {
-	// TODO(proofofkeags): Implement
-	return fn.None[FlushHookID]()
+func (l *channelLink) OnFlushedOnce(hook func()) fn.Option[FlushHookID] {
+	// If the channel state is clean, meaning there are no HTLCs, and both
+	// peers have irrevocably committed that state, then we can call this
+	// hook synchronously and not add it to the map at all.
+	if l.channel.IsChannelClean() {
+		hook()
+		return fn.None[FlushHookID]()
+	}
+
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	return fn.Some(FlushHookID(l.flushHooks.alloc(hook, false)))
 }
 
 // OnFlushedMany is a method that adds a hook to be called every time the
 // channel state reaches zero htlcs. This function returns a FlushId that can be
 // used to preemptively remove this hook.
-func (l *channelLink) OnFlushedMany(func()) FlushHookID {
-	// TODO(proofofkeags): Implement
-	return FlushHookID(0)
+func (l *channelLink) OnFlushedMany(hook func()) FlushHookID {
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	return FlushHookID(l.flushHooks.alloc(hook, true))
 }
 
 // RemoveFlushHook will remove a hook previously added by OnFlushedOnce or
 // OnFlushedMany. It returns an error if the supplied FlushId was not found.
-func (l *channelLink) RemoveFlushHook(FlushHookID) error {
-	// TODO(proofofkeags): Implement
-	return errors.New("no flush in progress to cancel")
+func (l *channelLink) RemoveFlushHook(hid FlushHookID) error {
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	if l.flushHooks.delete(uint64(hid)) {
+		return nil
+	}
+
+	return errors.New("could not remove flush hook: not found")
 }
 
 // OnCommitOnce adds a hook that will be called the next time a CommitSig
 // message is sent in the argument's LinkDirection.
 func (l *channelLink) OnCommitOnce(
-	LinkDirection, func(),
+	direction LinkDirection, hook func(),
 ) fn.Option[CommitHookID] {
 
-	// TODO(proofofkeags): Implement
-	return fn.None[CommitHookID]()
+	// If we want to run the hook once on the next fully committed state,
+	// and we don't currently need/owe a commitment, it means we don't have
+	// any pending updates and we can run the hook synchronously. If we call
+	// it syncrhonously we will not add it to the map at all.
+	runSync := direction == Outgoing && !l.channel.OweCommitment() ||
+		direction == Incoming && !l.channel.NeedCommitment()
+	if runSync {
+		hook()
```

Comment:
Ah ok I think this answers by question about calling the hook early (eg: to send shutdown as the coast is clear). I think would be useful to mention in the `godoc` comment for the method/interface, namely that if we're in a state where we don't actually need to send a commit, the hook will be called early (denoted by `None` hook ID).